cmd/collector/app/span_processor.go

```diff
@@ -177,3 +206,76 @@ func (sp *spanProcessor) enqueueSpan(span *model.Span, originalFormat SpanFormat
 	}
 	return sp.queue.Produce(item)
 }
+
+func (sp *spanProcessor) StartStateReporting(reportPeriod time.Duration) {
+	sp.background(reportPeriod, func() {
+		sp.updateGauges()
+	})
+}
+
+func (sp *spanProcessor) StartDynQueueSizeUpdater(reportPeriod time.Duration) {
+	sp.background(reportPeriod, func() {
+		sp.updateQueueSize()
+	})
+}
+
+func (sp *spanProcessor) background(reportPeriod time.Duration, callback func()) {
+	ticker := time.NewTicker(reportPeriod)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				callback()
+			case <-sp.stopCh:
+				return
+			}
+		}
+	}()
+}
+
+func (sp *spanProcessor) updateQueueSize() {
+	if sp.spansProcessed == 0 {
+		return
+	}
+
+	if sp.dynQueueSizeWarmup == 0 {
+		return
+	}
+
+	if sp.spansProcessed < uint64(sp.dynQueueSizeWarmup) {
+		return
+	}
+
+	// first, we get the average size of a span, by dividing the bytes processed by num of spans
+	average := sp.bytesProcessed / sp.spansProcessed
```

Comment:
it might make more sense to reset these variables once a resize happens, so that future resizes are based on recent data.

Reply:
Can we add that as something to improve in a future PR? There are some aspects that we should consider for this:

1. the effects of this as a metric (increase, increase, reset, increase, ...)
1. desirable/undesirable effects on quick bursts -- the way we have, quick bursts won't cause unnecessary resizes, which is a good thing if we are under heavy load (less processing), but if the burst is big enough, it might cause the collector to use more memory than it has available

Reply:
Created #2004 to track this